### PR TITLE
WIP - comparing external storage to grpc

### PR DESCRIPTION
**Pull request by marwan-at-work**
Things to note: 
We re-used the download protocol itself for all of the endpoints except save/delete. 

A super interesting fact: by adding those two endpoints to Athens, Athens itself can become an external storage server :meta:

Even more interesting, this would also enable direct uploads to Athens (registry style)

### CODE REVIEW

File: pkg/storage/external/server.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+		}
+		zip, err := strg.Zip(r.Context(), params.Module, params.Version)
+		if err != nil {
+			http.Error(w, err.Error(), errors.Kind(err))
+			return
+		}
+		defer zip.Close()
+		io.Copy(w, zip)
+	})
+	r.HandleFunc("/{module:.+}/@v/{version}.upload", func(w http.ResponseWriter, r *http.Request) {
```
Comment (arschles) [inline pkg/storage/external/server.go]: where is this called on the client? I see `.save` but not `.upload`

File: pkg/storage/external/client.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+			err = nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func (s *service) Save(ctx context.Context, module, version string, mod []byte, zip io.Reader, info []byte) error {
+	const op errors.Op = "external.Save"
+	url := s.url + "/" + module + "/@v/" + version + ".save"
```
Comment (arschles) [inline pkg/storage/external/client.go]: why not make this a `POST /module/@v/version`? same with deleting below when it gets implemented (`DELETE /module/@v/version`)

File: pkg/storage/external/server.go (hunk 1 above)
Comment (marwan-at-work) [reply pkg/storage/external/server.go]: Yep, that's a typo :) -- we should stick with either `.save` or `.upload` - any preference?

File: pkg/storage/external/client.go (hunk 2 above)
Comment (marwan-at-work) [reply pkg/storage/external/client.go]: @arschles i think it's better to be explicit and to take the `.ext` pattern. For example the download protocol explicitly does `GET /module/@v/list` and not just `GET /module` -- and

**Comment by marwan-at-work** [comment]
closing in favor of https://github.com/gomods/athens/pull/1587

**Pull request closed without merging**